waiter/src/waiter/scheduler/kubernetes.clj

```diff
@@ -1704,6 +1738,170 @@
                                   (assoc-in state [:rs-metadata :version :watch] version))))))}
       (merge options))))
 
+(defn- query-events
+  "Query K8s for all namespace-scoped events with matching object names. Returns a
+   channel containing the response or an exception."
+  [{:keys [api-server-url] :as scheduler} {:keys [api-request-fn]} {:keys [namespace k8s-object-name]}]
+  (log/info "fetching events from k8s for" namespace k8s-object-name)
+  (let [events-url (str api-server-url "/api/v1/namespaces/" namespace "/events?fieldSelector=involvedObject.name%3D" k8s-object-name)
+        events-chan (api-request-fn events-url scheduler)]
+    events-chan))
+
+(defn get-cached-events
+  "Load events from cache or query K8s and store the channel that will contain the response."
+  [k8s-object-key->event-cache k8s-object-key k8s-query-fn scheduler options]
+  (cu/cache-get-or-load k8s-object-key->event-cache k8s-object-key
+                        (fn load-event-from-k8s [] (k8s-query-fn scheduler options k8s-object-key))))
+
+(defn service-unhealthy?
+  "Return true if the given service appears unhealthy"
+  [{:keys [instances k8s/replicaset-creation-timestamp task-count task-stats]}]
+  (let [creation-timestamp (du/str-to-date-safe replicaset-creation-timestamp)
+        healthy-count (:healthy task-stats 0)
+        thirty-seconds-ago (t/ago (t/seconds 30))]
```

Comment:
Nitpick: we could make the 30 seconds configurable. Same below, we could use the same variable.